frontend/src/lib/query.js

```diff
@@ -390,6 +393,40 @@ var Query = {
         }
     },
 
+    addExpression(query) {
+        let expressions = query.expressions || {};
+        expressions[''] = [null, ''];
+        query.expressions = expressions;
+
+        console.log('added expressions. expressions =', query.expressions);
+    },
+
+    updateExpression(query, name, expression) {
+        query.expressions[name] = expression;
+
+        console.log('updated expression. expressions =', query.expressions);
+    },
+
+    renameExpression(query, oldName, newName) {
+        if (oldName === newName) return;
+
+        oldName = oldName || '';
+        query.expressions[newName] = query.expressions[oldName];
+        delete query.expressions[oldName];
+
+        console.log('renamed expression. expressions =', query.expressions);
+    },
+
+    removeExpression(query, name) {
+        if (!query.expressions) return;
+
+        delete query.expressions[name];
+
+        if (_.isEmpty(query.expressions)) delete query.expressions;
+
+        console.log('removed expression. expressions =', query.expressions);
+    },
```

Comment:
I don't have the final spec for what you intend the `:expressions` clause to look like, but the pattern that is forming is that `XYZClause` is meant to encapsulate all the changes within the clause and the only thing that should exist on `Query` is stuff that works on the entire clause such as `setExpressions(expressions)` would set the entire clause.

The functions you have such as `addExpression`, `updateExpression`, `renameExpression`, and `removeExpression` are all singular actions on the clause itself, so I think it's best if we separate those out in an `ExpressionClause` set of functions and unit test that on its own.